### PR TITLE
Refactor name and type resolution in preparation for free functions.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolidity/interface/CompilerStack.cpp [hunk 1]
```diff
@@ -326,28 +326,28 @@ bool CompilerStack::analyze()
...
-							solAssert(
-								m_errorReporter.hasErrors(),
-								"Contract already present (name clash?), but no error was reported."
-							);
-					}
-
+					// Note that we now reference contracts by their fully qualified names, and
+					// thus contracts can only conflict if declared in the same source file. This
+					// should already cause a double-declaration error elsewhere.
+					if (m_contracts.find(contract->fullyQualifiedName()) == m_contracts.end())
```
Comment (leonardoalt) [inline libsolidity/interface/CompilerStack.cpp]: Might be a matter of preference, but I prefer
`!m_contracts.count(...)`

**Comment by chriseth** [comment]
Updated.

File: libsolidity/analysis/NameAndTypeResolver.cpp [hunk 2]
```diff
@@ -227,13 +230,14 @@ bool NameAndTypeResolver::resolveNamesAndTypesInternal(ASTNode& _node, bool _res
...
 		solAssert(!!m_currentScope, "");
+		solAssert(_resolveInsideCode, "");
 
 		m_globalContext.setCurrentContract(*contract);
 		updateDeclaration(*m_globalContext.currentSuper());
 		updateDeclaration(*m_globalContext.currentThis());
 
 		for (ASTPointer<InheritanceSpecifier> const& baseContract: contract->baseContracts())
-			if (!resolveNamesAndTypes(*baseContract, true))
+			if (!resolveNamesAndTypesInternal(*baseContract, true))
```
Comment (hrkrshnn) [inline libsolidity/analysis/NameAndTypeResolver.cpp:240]: This won't be wrapped inside that try-catch block anymore. Will that change anything?

Comment (hrkrshnn) [reply libsolidity/analysis/NameAndTypeResolver.cpp:240]: Ok. This won't make any difference.

**Review by hrkrshnn — APPROVED** [review]
LGTM!

**Pull request merged**